chips/lowrisc/src/gpio.rs

```diff
@@ -0,0 +1,252 @@
+    direct_oe: ReadWrite<u32, pins::Register>,
+    // GPIO write Output Enable lower with mask
+    masked_oe_low: ReadWrite<u32, pins::Register>,
+    // GPIO write Output Enable upper with mask
+    masked_oe_upper: ReadWrite<u32, pins::Register>,
+    // GPIO interrupt enable for GPIO, rising edge
+    intr_ctrl_en_rise: ReadWrite<u32, pins::Register>,
+    // GPIO interrupt enable for GPIO, falling edge
+    intr_ctrl_en_fall: ReadWrite<u32, pins::Register>,
+    // GPIO interrupt enable for GPIO, level high
+    intr_ctrl_en_lvlhigh: ReadWrite<u32, pins::Register>,
+    // GPIO interrupt enable for GPIO, level low
+    intr_ctrl_en_lvllow: ReadWrite<u32, pins::Register>,
+    // filter enable for GPIO input bits
+    intr_ctrl_en_input_filer: ReadWrite<u32, pins::Register>,
+}
+
+register_bitfields![u32,
+	pins [
+	    pin0 0,
+	    pin1 1,
+	    pin2 2,
+	    pin3 3,
+	    pin4 4,
+	    pin5 5,
+	    pin6 6,
+	    pin7 7,
+	    pin8 8,
+	    pin9 9,
+	    pin10 10,
+	    pin11 11,
+	    pin12 12,
+	    pin13 13,
+	    pin14 14,
+	    pin15 15,
+	    pin16 16,
+	    pin17 17,
+	    pin18 18,
+	    pin19 19,
+	    pin20 20,
+	    pin21 21,
+	    pin22 22,
+	    pin23 23,
+	    pin24 24,
+	    pin25 25,
+	    pin26 26,
+	    pin27 27,
+	    pin28 28,
+	    pin29 29,
+	    pin30 30,
+	    pin31 31
+	]
+];
+
+pub struct GpioPin {
+    registers: StaticRef<GpioRegisters>,
+    pin: Field<u32, pins::Register>,
+    set: FieldValue<u32, pins::Register>,
+    clear: FieldValue<u32, pins::Register>,
+    client: OptionalCell<&'static dyn hil::gpio::Client>,
+}
+
+impl GpioPin {
+    pub const fn new(
+        base: StaticRef<GpioRegisters>,
+        pin: Field<u32, pins::Register>,
+        set: FieldValue<u32, pins::Register>,
+        clear: FieldValue<u32, pins::Register>,
+    ) -> GpioPin {
+        GpioPin {
+            registers: base,
+            pin: pin,
+            set: set,
+            clear: clear,
+            client: OptionalCell::empty(),
+        }
+    }
+
+    /// There are separate interrupts in PLIC for each pin, so the interrupt
+    /// handler only needs to exist on each pin.
+    pub fn handle_interrupt(&self) {
+        // TODO
+    }
+}
+
+impl hil::gpio::Configure for GpioPin {
+    fn configuration(&self) -> hil::gpio::Configuration {
+        let regs = self.registers;
+
+        let output = regs.direct_oe.is_set(self.pin);
+        let input;
+        if self.pin.shift < 15 {
+            input = regs.masked_oe_low.is_set(self.pin);
+        } else {
+            input = regs.masked_oe_upper.is_set(self.pin);
+        }
+
+        return match (input, output) {
+            (true, true) => hil::gpio::Configuration::InputOutput,
+            (true, false) => hil::gpio::Configuration::Input,
+            (false, true) => hil::gpio::Configuration::Output,
+            (false, false) => hil::gpio::Configuration::LowPower,
+        };
+    }
+
+    fn set_floating_state(&self, _mode: hil::gpio::FloatingState) {
```

Comment:
Can this be implemented?

Reply:
I don't see how with the current registers: https://github.com/lowRISC/opentitan/blob/master/hw/ip/gpio/data/gpio.hjson